File: swift/Sources/Thoaded/Thoaded.swift
/* The source for a library for simple and fast design pattern of model,
view and controller in supported modern programming languages.
Author: Walber Antonio Ramos Beltrame (walber.beltrame@gmail.com) */

// The generic protocol represents a task events.
protocol Task {}

// The generic protocol represents a single dispatcher for all modifying events.
protocol Modified {}

// The generic protocol represents a list events.
protocol Thoaded {

    associatedtype T

    // The event represents a task of adding some object.
    func added(_ t: T) -> Task

    // The event represents a task of updating some object.
    func updated(_ t: T) -> Task

    // The event represents a task of deleting some object.
    func deleted(_ t: T) -> Task

    // The event represents a task of reading some object.
    func readed(_ t: T) -> Task

    // The event represents a task of working some object.
    func worked(_ t: T) -> Task

    // The event represents a task of handling some object.
    func handled(_ t: T) -> Task

    // The event represents a task of querying objects.
    func queried(_ t: T...) -> [Task]

    // The event represents a task of listening some object.
    func listened(_ t: T...) -> Task

    // The event represents a task of unlistening some object.
    func unlistened(_ t: T...) -> Task

}

// A modeled protocol might have a single listener for all model events in source.
protocol Modeled : Thoaded {}

// A viewed protocol might have a single listener for all view events in source.
protocol Viewed : Thoaded {}

// The generic protocol represents a control of events.
protocol Controlled {}

File: java/src/main/java/com/walberbeltrame/thoaded/Thoaded.java
/**
 * 
 * The open source initiative for a library for simple and fast design pattern that extends model,
 * view and controller in supported modern programming languages.
 * 
 * @author Walber Antonio Ramos Beltrame (walber.beltrame@gmail.com)
 * 
 */
package com.walberbeltrame.thoaded;

import java.util.concurrent.Future;

/**
 * 
 * The generic class represents a list events.
 * 
 * @param <T>
 */
public abstract class Thoaded<T> {

 private Thoaded<T> modified;

 /**
  * 
  * @return The thoaded object to a single dispatcher for all modifying events.
  */
 public Thoaded<T> getModified() {
  return this.modified;
 }

 /**
  * 
  * @param modified
  */
 public void setModified(Thoaded<T> modified) {
  this.modified = modified;
 }

 /**
  * 
  * @param t T
  * @return The event represents a future of adding some object.
  */
 public abstract Future<T> added(T t);

 /**
  * 
  * @param t T
  * @return The event represents a future of updating some object.
  */
 public abstract Future<T> updated(T t);

 /**
  * 
  * @param t T
  * @return The event represents a future of deleting some object.
  */
 public abstract Future<T> deleted(T t);

 /**
  * 
  * @param t T
  * @return The event represents a future of reading some object.
  */
 public abstract Future<T> readed(T t);

 /**
  * 
  * @param t T
  * @return The event represents a future of working some object.
  */
  public abstract Future<T> worked(T t);

  /**
   * 
   * @param t T
   * @return The event represents a future of handling some object.
   */
  public abstract Future<T> handled(T t);

 /**
  * 
  * @param t T
  * @return The event represents a future of querying objects.
  */
 public abstract Future<Iterable<T>> queried(T[] t);

 /**
  * 
  * @param t T
  * @return The event represents a future of listening some object.
  */
 public abstract Future<T> listened(T[] t);
 
 /**
  * 
  * @param t T
  * @return The event represents a future of unlistening some object.
  */
 public abstract Future<T> unlistened(T[] t);

}